lib/shopify_app/controller_concerns/payload_verification.rb

```diff
@@ -0,0 +1,43 @@
+# frozen_string_literal: true
+module ShopifyApp
+  module PayloadVerification
+    extend ActiveSupport::Concern
+
+    def hmac_valid?(data)
+      secrets = [ShopifyApp.configuration.secret, ShopifyApp.configuration.old_secret].reject(&:blank?)
+
+      secrets.any? do |secret|
+        digest = OpenSSL::Digest.new('sha256')
+
+        ActiveSupport::SecurityUtils.secure_compare(
+          shopify_hmac,
+          Base64.strict_encode64(OpenSSL::HMAC.digest(digest, secret, data))
+        )
+      end
+    end
+
+    def shopify_hmac
+      request.headers['HTTP_X_SHOPIFY_HMAC_SHA256']
+    end
+
+    def verify_request
+      hmac_header = request.headers['HTTP_X_SHOPIFY_HMAC_SHA256']
+      request_body = request.body.read
+      secret = ShopifyApp.configuration.secret
+      digest = OpenSSL::Digest.new('sha256')
+
+      expected_hmac = Base64.strict_encode64(OpenSSL::HMAC.digest(digest, secret, request_body))
+      head(:unauthorized) unless ActiveSupport::SecurityUtils.secure_compare(expected_hmac, hmac_header)
+    end
+
+    def calculated_signature(query_hash_without_signature)
+      sorted_params = query_hash_without_signature.collect { |k, v| "#{k}=#{Array(v).join(',')}" }.sort.join
+
+      OpenSSL::HMAC.hexdigest(
+        OpenSSL::Digest.new('sha256'),
+        ShopifyApp.configuration.secret,
+        sorted_params
+      )
+    end
```

Comment:
There is a lot of repitition on this and I think it could be simpler. These two methods are really just the following.

```suggestion
    def verify_request
      head(:unauthorized) unless hmac_valid?(request.body.read) 
    end

    def calculated_signature(query_hash_without_signature)
      sorted_params = query_hash_without_signature.collect { |k, v| "#{k}=#{Array(v).join(',')}" }.sort.join
      hmac_valid?(sorted_params)
    end
```

The simplified versions of `verify_request` and `calculated_signature` could be put back where they came from because they don't need to be shared around and then other code can simply leverage the `hmac_valid?` method which means we have allowed for secret fallback in these two extra instances.

Reply:
@tanema, 

Thanks for the feedback, I was able to refactor the method `verify_request` and move the logic back in `app/controllers/shopify_app/extension_verification_controller.rb`. But the `calculated_signiture` does not work with the refactor for two reasons. The reasons are: 
1. Hmac in the `calculated_signiture` looks for app-level secrets, whereas the Hmac in `hmac_valid?` looks for secrets at the request level. 
2. Hmac in `calculated_signiture` returns a string, but `hmac_valid?` returns boolean.  Because of these reasons, the refactor didn't work. and I have put the code back to what it was before the refactor. Please let me know how you wanted to proceed.